w3af_batch

```diff
@@ -0,0 +1,136 @@
+import multiprocessing as mp
+import multiprocessing.dummy as mpd
+
+
+WAIT_TIMEOUT = 30  # in seconds
+STOPSCAN = mp.Event()
+logger = logging.getLogger('w3af_batch')
+
+
+def cancel_scan(*args):
+    """
+    Dummy SIGINT hanlder which stops target list processing.
+    """
+    STOPSCAN.set()
+    raise KeyboardInterrupt('Scan is terminated by user request')
+
+
+def execute_scan(target, profile, timeout):
+    """
+    Execute w3af scan. Try to stop it after timeout seconds.
+    """
+    def stop_scan(scanner):
+        """
+        Stop w3af scan after timeout.
+        """
+        if scanner.status.is_running():
+            scanner.stop()
+            logger.info('Stop scan of %s because of configured timeout',
+                        target)
+
+    from w3af.core.controllers.w3afCore import w3afCore
+    from w3af.core.data.kb.knowledge_base import kb
+
+    scanner = w3afCore()
+    scanner.profiles.use_profile(profile)
+    option_list = scanner.target.get_options()
+    option_list['target'].set_value(target)
+    scanner.target.set_options(option_list)
+    scanner.plugins.init_plugins()
+
+    timer = th.Timer(timeout, stop_scan, (scanner,))
+    timer.start()
+    scanner.start()
+    if timer.is_alive():
+        timer.cancel()
+        timer.join()
+        logger.debug('Target %s has %s vulnerabilities',
+                     target, len(kb.get_all_vulns()))
+
+
+def start_executor(target, profile, timeout):
+    """
+    Start w3af scan in a separate process. Terminate child process after
+    WAIT_TIMEOUT seconds.
+
+    :param target: Target uri
+    :param profile: Profile's internal name of path to pw3af file
+    :param timeout: Scan timeout
+    """
+    def stop_executor(executor, target):
+        """
+        Force stop of w3af scanner process after WAIT_TIMEOUT.
+        """
+        if executor.is_alive():
+            executor.terminate()
+            logger.info('Terminate scan proccess of %s '
+                        'after waiting for w3afCore.stop()',
+                        target)
+
+    executor = mp.Process(target=execute_scan,
+                          name='Executor',
+                          args=(target, profile, timeout))
+    timer = th.Timer(timeout + WAIT_TIMEOUT, stop_executor, (executor, target))
+    timer.start()
+    executor.start()
+    executor.join()
+    if timer.is_alive():
+        timer.cancel()
+        timer.join()
+
+
+def main(args):
+    """
+    Execute w3af scans of args.targets in parallel.  A pool of manager
+    threads is created. Each thread receives a target to scan. Manager
+    thread executes scan in a separate process. Scanner process will
+    try to stop w3af after timeout seconds. In addition manager thread
+    will terminate the child process after WAIT_TIMEOUT seconds.
+    """
+    def terminate_scan(pool):
+        """
+        Terminate processing of pool tasks in case of SIGINT.
+        """
+        if STOPSCAN.wait():
+            pool.terminate()
```

Comment:
The Ctrl+C handling "timeout" might be hanging in this line, because `pool.terminate()` doesn't kill the pool threads, it just clears the task queue and calls join() on the running threads. To make sure that it all exits quickly you might want to use the STOPSCAN to (also) call `.terminate()` on the running executor(s)